### ISSUE TITLE
intermittent "ClassName is incompatible with ClassName"

### ISSUE THREAD

**Issue by matejcik** [microsoft/pylance-release#4151]
Type: <b>Bug</b>

1. create the following two files:

```
# base.py
from typing import Generic, TypeVar
from typing_extensions import dataclass_transform

T = TypeVar("T")

@dataclass_transform()
class Base:
    pass

class Derived(Base, Generic[T]):
    pass
```
and
```
# poc.py
from base import Derived


class Example(Derived):
    x: int

ex = Example(x=1)
```

2. open both files in tabs

3. modify the definition of `Base` class, e.g., place a newline between the class and "pass"

### Expected result:

Nothing interesting.

### Actual result:

The `poc.py` tab goes red.

the line `ex = Example(x=1)` has red underline and the following pylance error:
```
Could not bind method "__new__" because "Type[Example]" is not assignable to parameter "cls"
  "poc.Example" is incompatible with "poc.Example" (PylancereportGeneralTypeIssues)
```

the error disappears after (a) reverting the change in base.py, or (b) after any modification in poc.py

Extension version: 2023.3.30
VS Code version: Code 1.76.2 (ee2b180d582a7f601fa6ecfdad8d9fd269ab1884, 2023-03-14T17:57:21.103Z)
OS version: Linux x64 6.0.19-060019-generic
Modes:
Sandboxed: Yes

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 5800H with Radeon Graphics (16 x 4440)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|1, 1, 1|
|Memory (System)|15.02GB (5.72GB free)|
|Process Argv|--enable-features=UseOzonePlatform --ozone-platform=wayland . --enable-proposed-api genuitecllc.codetogether --crash-reporter-id aab2c7ab-1587-4f4d-882e-b3efef3177c3|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|gnome|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|gnome|
|XDG_SESSION_TYPE|wayland|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593cf:30376535
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
pyindex848:30662994
nodejswelcome1:30587005
3biah626:30602489
pyind779:30671433
f6dab269:30613381
pythonsymbol12:30671437
a9j8j154:30646983
showlangstatbar:30672706
vsctsb:30677850
azdwalk:30687957
pythonms35:30686772

```

</details>

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in an upcoming version of pylance.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023341-29-March-2023-PreRelease)
